### ISSUE TITLE
Does not play nice with Factory Girl

### ISSUE THREAD

**Issue by muzfuz** [ruby-analysis/delfos#14]
The way Factory Girl associations work means that Delfos will try to log method calls.

This quick hack makes it stop, but we need to actually intervene much earlier as we're not really interested in anything that FactoryGirl is doing.

```ruby
    def setup_trace_point(type, klass)
        TracePoint.new(type) do |tp|
          next unless AppDirectories.include_files?(tp.path)

          ### HACK
          require "binding_of_caller"
          c = tp.binding.of_caller(2).eval("self.class")
          next if c.is_a? FactoryGirl::Declaration::Implicit
          next if c == FactoryGirl::AttributeAssigner
          ### END HACK

          klass.new(tp).perform
        end
      end
```

Opening this issue now so we can come back to it later.

**Comment by markburns** [comment]
Obvious solution is allow excluding directories and/or using a regex/globmatch.
Actually in the past I think I had a regex match but removed it for simplicity's sake.
Happy to accept that as a change.

Also though, in your specific case do you need to use the spec directory at all?

**Comment by markburns** [comment]
Or perhaps you mean factory girl is currently getting logged and shouldn't be?
If so and the method definitions are outside of Delfos.application_directories then it's probably a bug

**Comment by muzfuz** [comment]
Scenario is:

- We define `app` as an application directory.
- Factory Girl creates `ObjectA`, with an association on `ObjectB`.
- Delfos tries to do a method trace on `ObjectB` and then blows up.

**Comment by markburns** [comment]
Oh right. So it's probably more serious than just FactoryGirl it's more like
App code -> third party library -> app code that causes it.
Sounds easy enough to replicate in a test

**Comment by muzfuz** [comment]
Having a closer look at this now.

**Comment by muzfuz** [comment]
![screen shot 2017-06-13 at 16 27 54](https://user-images.githubusercontent.com/7075351/27090513-47ecca86-5055-11e7-8bb8-a49676a903bd.png)

Here's what that spec has generated -- looks like it's failing in the way we expected it to.

**Comment by markburns** [comment]
I guess I'd expect it to behave like this instead of one call stack like this.

```
app_code.rb:1
app_code.rb:3
lib_code.rb:5
app_code.rb:7
app_code.rb:10
```

We should get two separate call stacks like this
```
app_code.rb:1
app_code.rb:3
```

```
app_code.rb:7
app_code.rb:10
```

**Comment by markburns** [comment]
@muzfuz - have a spec you might want to use if you're looking at it in #18

**Comment by markburns** [comment]
Fixed in #19
